src/Agent.Worker/Build/ArtifactCommands.cs

```diff
@@ -0,0 +1,339 @@
+                                                         artifactType,
+                                                         artifactData,
+                                                         propertyDictionary,
+                                                         context.CancellationToken);
+            context.AsyncCommands.Add(commandContext);
+        }
+
+        private void ProcessArtifactUploadCommand(IExecutionContext context, Dictionary<string, string> eventProperties, string data)
+        {
+            ArgUtil.NotNull(context, nameof(context));
+            ArgUtil.NotNull(context.Endpoints, nameof(context.Endpoints));
+
+            ServiceEndpoint systemConnection = context.Endpoints.FirstOrDefault(e => string.Equals(e.Name, ServiceEndpoints.SystemVssConnection, StringComparison.OrdinalIgnoreCase));
+            ArgUtil.NotNull(systemConnection, nameof(systemConnection));
+            ArgUtil.NotNull(systemConnection.Url, nameof(systemConnection.Url));
+
+            Uri projectUrl = systemConnection.Url;
+            VssCredentials projectCredential = ApiUtil.GetVssCredential(systemConnection);
+
+            Guid projectId = context.Variables.System_TeamProjectId ?? Guid.Empty;
+            ArgUtil.NotEmpty(projectId, nameof(projectId));
+
+            int? buildId = context.Variables.Build_BuildId;
+            ArgUtil.NotNull(buildId, nameof(buildId));
+
+            long? containerId = context.Variables.Build_ContainerId;
+            ArgUtil.NotNull(containerId, nameof(containerId));
+
+            string artifactName;
+            if (!eventProperties.TryGetValue(ArtifactAssociateEventProperties.ArtifactName, out artifactName))
+            {
+                throw new Exception(StringUtil.Loc("ArtifactNameRequired"));
+            }
+
+            string containerFolder;
+            if (!eventProperties.TryGetValue(ArtifactUploadEventProperties.ContainerFolder, out containerFolder))
+            {
+                containerFolder = artifactName;
+            }
+
+            var propertyDictionary = ExtractArtifactProperties(eventProperties);
+
+            string localPath = data;
+            if (string.IsNullOrEmpty(localPath))
+            {
+                throw new Exception(StringUtil.Loc("ArtifactLocationRequired"));
+            }
+
+            string fullPath = Path.GetFullPath(localPath);
+            if (!File.Exists(fullPath) && !Directory.Exists(fullPath))
+            {
+                // if localPath is not a file or folder on disk
+                throw new FileNotFoundException(StringUtil.Loc("PathNotExist", localPath));
+            }
+            else if (Directory.Exists(fullPath) && Directory.GetFiles(fullPath, "*", SearchOption.AllDirectories).Length == 0)
+            {
+                // if localPath is a folder but the folder contains nothing
+                context.Warning(StringUtil.Loc("DirectoryIsEmpty", fullPath));
```

Comment:
I can't figure out whether a more descriptive error message would be better? If the end user doesn't find their artifact on the build summary page, are they going to intuitively connect the dots that it's because "Directory: {0} is empty"? I'm inclined to think it will not always be intuitive. If more context would be helpful, we even know the artifact name here at this point.

I'm not 100% sure that my leaning is correct. I just put the question out there, you decide.